File: src-tauri/src/analytics/productivity.rs
// Productivity metrics — per-day KPIs for sessions started/completed, active
// minutes, tool calls, and token distribution percentiles.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use chrono::{Datelike, NaiveDate};
use serde::{Deserialize, Serialize};

use crate::discovery::{path_decoder, project_scanner, subproject_registry::SubprojectRegistry};
use crate::watcher;

use super::session_scan::scan_session_fast;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProductivityDay {
    pub date: String,
    pub sessions_started: u32,
    pub sessions_completed: u32,
    pub active_ms: f64,
    pub tool_calls: u64,
    pub tokens_p50: u64,
    pub tokens_p95: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProductivityTotals {
    pub sessions_started: u32,
    pub sessions_completed: u32,
    pub active_ms: f64,
    pub tool_calls: u64,
    pub tokens_p50: u64,
    pub tokens_p95: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProductivityMetrics {
    pub days: Vec<ProductivityDay>,
    pub totals: ProductivityTotals,
}

/// Compute percentile of a sorted ascending `u64` slice. Empty → 0.
pub fn percentile_u64(sorted: &[u64], pct: f64) -> u64 {
    if sorted.is_empty() {
        return 0;
    }
    let p = pct.clamp(0.0, 1.0);
    let len = sorted.len();
    let idx = ((len as f64 - 1.0) * p).round() as usize;
    sorted[idx.min(len - 1)]
}

fn day_key_ms(ms: f64) -> Option<String> {
    let secs = (ms / 1000.0) as i64;
    chrono::DateTime::from_timestamp(secs, 0)
        .map(|dt| format!("{:04}-{:02}-{:02}", dt.year(), dt.month(), dt.day()))
}

pub fn compute_productivity_metrics(
    days: u32,
    registry: &Arc<Mutex<SubprojectRegistry>>,
) -> Result<ProductivityMetrics, String> {
    let claude_dir = watcher::resolve_claude_dir().ok_or("Cannot resolve home directory")?;
    let projects_dir = path_decoder::get_projects_base_path(&claude_dir);

    let days = days.clamp(1, 90);

    let projects = {
        let mut reg = registry.lock().map_err(|e| e.to_string())?;
        project_scanner::scan_projects(&projects_dir, &mut reg)?
    };

    let now_ms = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as f64;
    let cutoff_ms = now_ms - (days as f64) * 86_400_000.0;

    // Seed empty day buckets for the requested range so sparklines have a full
    // timeline even if some days are idle.
    let mut day_buckets: HashMap<String, DayAccumulator> = HashMap::new();
    let now_secs = (now_ms / 1000.0) as i64;
    let today = chrono::DateTime::from_timestamp(now_secs, 0)
        .map(|dt| dt.date_naive())
        .unwrap_or_else(|| NaiveDate::from_ymd_opt(1970, 1, 1).unwrap_or_default());
    let mut ordered_keys: Vec<String> = Vec::with_capacity(days as usize);
    for i in (0..days as i64).rev() {
        let d = today - chrono::Duration::days(i);
        let key = format!("{:04}-{:02}-{:02}", d.year(), d.month(), d.day());
        ordered_keys.push(key.clone());
        day_buckets.entry(key).or_default();
    }

    let mut seen_dirs = std::collections::HashSet::new();

    for project in &projects {
        let base_id = match project.id.find("::") {
            Some(idx) => &project.id[..idx],
            None => &project.id,
        };
        if !seen_dirs.insert(base_id.to_string()) {
            continue;
        }

        let project_dir = projects_dir.join(base_id);
        if !project_dir.is_dir() {
            continue;
        }

        let entries = match std::fs::read_dir(&project_dir) {
            Ok(e) => e,
            Err(_) => continue,
        };

        for entry in entries.flatten() {
            let fname = entry.file_name();
            let fname = fname.to_string_lossy();
            if !fname.ends_with(".jsonl") {
                continue;
            }

            let modified_at = entry
                .metadata()
                .ok()
                .and_then(|m| m.modified().ok())
                .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                .map(|d| d.as_secs_f64() * 1000.0)
                .unwrap_or(0.0);
            let created_at = entry
                .metadata()
                .ok()
                .and_then(|m| m.created().ok())
                .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                .map(|d| d.as_secs_f64() * 1000.0)
                .unwrap_or(0.0);
            let latest = modified_at.max(created_at);
            if latest < cutoff_ms {
                continue;
            }

            let summary = match scan_session_fast(&entry.path()) {
                Some(s) => s,
                None => continue,
            };

            let start_ms = summary.first_timestamp_ms.unwrap_or(created_at);
            let end_ms = summary.last_timestamp_ms.unwrap_or(latest);

            let started_key = day_key_ms(start_ms);
            let completed_key = if summary.duration_ms > 0.0 {
                day_key_ms(end_ms)
            } else {
                None
            };

            if let Some(key) = started_key {
                let bucket = day_buckets.entry(key).or_default();
                bucket.sessions_started += 1;
                bucket.active_ms += summary.active_ms;
                bucket.tool_calls += summary.tool_call_count;
                let tokens = summary.input_tokens
                    + summary.output_tokens
                    + summary.cache_read_tokens
                    + summary.cache_creation_tokens;
                bucket.tokens.push(tokens);
            }

            if let Some(key) = completed_key {
                let bucket = day_buckets.entry(key).or_default();
                bucket.sessions_completed += 1;
            }
        }
    }

    // Build the days array in chronological order (oldest → newest).
    let mut days_out: Vec<ProductivityDay> = Vec::with_capacity(ordered_keys.len());
    let mut all_tokens: Vec<u64> = Vec::new();
    let mut totals = ProductivityTotals {
        sessions_started: 0,
        sessions_completed: 0,
        active_ms: 0.0,
        tool_calls: 0,
        tokens_p50: 0,
        tokens_p95: 0,
    };

    for key in ordered_keys {
        let bucket = day_buckets.remove(&key).unwrap_or_default();
        let mut sorted = bucket.tokens.clone();
        sorted.sort_unstable();
        let tokens_p50 = percentile_u64(&sorted, 0.5);
        let tokens_p95 = percentile_u64(&sorted, 0.95);

        totals.sessions_started += bucket.sessions_started;
        totals.sessions_completed += bucket.sessions_completed;
        totals.active_ms += bucket.active_ms;
        totals.tool_calls += bucket.tool_calls;
        all_tokens.extend_from_slice(&sorted);

        days_out.push(ProductivityDay {
            date: key,
            sessions_started: bucket.sessions_started,
            sessions_completed: bucket.sessions_completed,
            active_ms: bucket.active_ms,
            tool_calls: bucket.tool_calls,
            tokens_p50,
            tokens_p95,
        });
    }

    all_tokens.sort_unstable();
    totals.tokens_p50 = percentile_u64(&all_tokens, 0.5);
    totals.tokens_p95 = percentile_u64(&all_tokens, 0.95);

    Ok(ProductivityMetrics {
        days: days_out,
        totals,
    })
}

#[derive(Default)]
struct DayAccumulator {
    sessions_started: u32,
    sessions_completed: u32,
    active_ms: f64,
    tool_calls: u64,
    tokens: Vec<u64>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn percentile_empty_zero() {
        assert_eq!(percentile_u64(&[], 0.5), 0);
    }

    #[test]
    fn percentile_median() {
        let data: Vec<u64> = vec![1, 2, 3, 4, 5];
        assert_eq!(percentile_u64(&data, 0.5), 3);
    }

    #[test]
    fn percentile_p95_small_sample_rounds() {
        let data: Vec<u64> = vec![10, 20, 30, 40, 50];
        assert_eq!(percentile_u64(&data, 0.95), 50);
    }

    #[test]
    fn percentile_clamps_upper() {
        let data: Vec<u64> = vec![7];
        assert_eq!(percentile_u64(&data, 0.99), 7);
    }
}

File: internal/analytics/productivity.go
// Productivity metrics — mirrors src-tauri/src/analytics/productivity.rs.
package analytics

import (
	"fmt"
	"os"
	"sort"
	"strings"
	"time"

	"claude-devtools/internal/discovery"
)

// ProductivityDay holds per-day KPIs.
type ProductivityDay struct {
	Date               string  `json:"date"`
	SessionsStarted    uint32  `json:"sessionsStarted"`
	SessionsCompleted  uint32  `json:"sessionsCompleted"`
	ActiveMs           float64 `json:"activeMs"`
	ToolCalls          uint64  `json:"toolCalls"`
	TokensP50          uint64  `json:"tokensP50"`
	TokensP95          uint64  `json:"tokensP95"`
}

// ProductivityTotals holds aggregated KPIs across all days.
type ProductivityTotals struct {
	SessionsStarted   uint32  `json:"sessionsStarted"`
	SessionsCompleted uint32  `json:"sessionsCompleted"`
	ActiveMs          float64 `json:"activeMs"`
	ToolCalls         uint64  `json:"toolCalls"`
	TokensP50         uint64  `json:"tokensP50"`
	TokensP95         uint64  `json:"tokensP95"`
}

// ProductivityMetrics is the payload returned by GetProductivityMetrics.
type ProductivityMetrics struct {
	Days   []ProductivityDay  `json:"days"`
	Totals ProductivityTotals `json:"totals"`
}

// PercentileU64 returns the p-th percentile of a sorted ascending uint64 slice.
// Empty → 0. Mirrors productivity::percentile_u64.
func PercentileU64(sorted []uint64, pct float64) uint64 {
	if len(sorted) == 0 {
		return 0
	}
	if pct < 0.0 {
		pct = 0.0
	}
	if pct > 1.0 {
		pct = 1.0
	}
	idx := int((float64(len(sorted)-1)*pct)+0.5) // round
	if idx >= len(sorted) {
		idx = len(sorted) - 1
	}
	return sorted[idx]
}

type dayAccumulator struct {
	sessionsStarted   uint32
	sessionsCompleted uint32
	activeMs          float64
	toolCalls         uint64
	tokens            []uint64
}

func dayKeyMs(ms float64) string {
	t := msToTime(ms)
	return fmt.Sprintf("%04d-%02d-%02d", t.Year(), int(t.Month()), t.Day())
}

// ComputeProductivityMetrics returns per-day productivity KPIs for the last `days` days.
// Mirrors productivity::compute_productivity_metrics.
func ComputeProductivityMetrics(days uint32) (*ProductivityMetrics, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil, fmt.Errorf("cannot resolve home directory: %w", err)
	}
	projectsDir := discovery.GetProjectsBasePath(home + "/.claude")

	if days < 1 {
		days = 1
	}
	if days > 90 {
		days = 90
	}

	registry := discovery.NewSubprojectRegistry()
	projects, err := discovery.ScanProjects(projectsDir, registry)
	if err != nil {
		return nil, fmt.Errorf("scan projects: %w", err)
	}

	nowMs := float64(time.Now().UnixMilli())
	cutoffMs := nowMs - float64(days)*86_400_000.0

	// Seed empty day buckets for the full range (oldest → newest).
	today := msToTime(nowMs).Truncate(24 * time.Hour)
	orderedKeys := make([]string, 0, days)
	dayBuckets := map[string]*dayAccumulator{}
	for i := int(days) - 1; i >= 0; i-- {
		d := today.AddDate(0, 0, -i)
		key := fmt.Sprintf("%04d-%02d-%02d", d.Year(), int(d.Month()), d.Day())
		orderedKeys = append(orderedKeys, key)
		dayBuckets[key] = &dayAccumulator{}
	}

	seenDirs := map[string]struct{}{}

	for _, project := range projects {
		baseID := discovery.ExtractBaseDir(project.ID)
		if _, ok := seenDirs[baseID]; ok {
			continue
		}
		seenDirs[baseID] = struct{}{}

		projectDir := projectsDir + "/" + baseID
		info, err := os.Stat(projectDir)
		if err != nil || !info.IsDir() {
			continue
		}

		entries, err := os.ReadDir(projectDir)
		if err != nil {
			continue
		}

		for _, entry := range entries {
			name := entry.Name()
			if entry.IsDir() || !strings.HasSuffix(name, ".jsonl") {
				continue
			}

			fi, err := entry.Info()
			if err != nil {
				continue
			}
			modifiedMs := float64(fi.ModTime().UnixMilli())
			createdMs := modifiedMs // no portable birthtime in Go; use mtime

			latest := modifiedMs
			if createdMs > latest {
				latest = createdMs
			}
			if latest < cutoffMs {
				continue
			}

			summary := ScanSessionFast(projectDir + "/" + name)
			if summary == nil {
				continue
			}

			startMs := createdMs
			if summary.FirstTimestampMs != nil {
				startMs = *summary.FirstTimestampMs
			}
			endMs := latest
			if summary.LastTimestampMs != nil {
				endMs = *summary.LastTimestampMs
			}

			tokens := summary.InputTokens + summary.OutputTokens +
				summary.CacheReadTokens + summary.CacheCreationTokens

			startedKey := dayKeyMs(startMs)
			if bucket, ok := dayBuckets[startedKey]; ok {
				bucket.sessionsStarted++
				bucket.activeMs += summary.ActiveMs
				bucket.toolCalls += summary.ToolCallCount
				bucket.tokens = append(bucket.tokens, tokens)
			}

			if summary.DurationMs > 0.0 {
				completedKey := dayKeyMs(endMs)
				if bucket, ok := dayBuckets[completedKey]; ok {
					bucket.sessionsCompleted++
				}
			}
		}
	}

	daysOut := make([]ProductivityDay, 0, len(orderedKeys))
	var allTokens []uint64
	totals := ProductivityTotals{}

	for _, key := range orderedKeys {
		bucket := dayBuckets[key]
		sorted := make([]uint64, len(bucket.tokens))
		copy(sorted, bucket.tokens)
		sort.Slice(sorted, func(i, j int) bool { return sorted[i] < sorted[j] })
		p50 := PercentileU64(sorted, 0.5)
		p95 := PercentileU64(sorted, 0.95)

		totals.SessionsStarted += bucket.sessionsStarted
		totals.SessionsCompleted += bucket.sessionsCompleted
		totals.ActiveMs += bucket.activeMs
		totals.ToolCalls += bucket.toolCalls
		allTokens = append(allTokens, sorted...)

		daysOut = append(daysOut, ProductivityDay{
			Date:              key,
			SessionsStarted:   bucket.sessionsStarted,
			SessionsCompleted: bucket.sessionsCompleted,
			ActiveMs:          bucket.activeMs,
			ToolCalls:         bucket.toolCalls,
			TokensP50:         p50,
			TokensP95:         p95,
		})
	}

	sort.Slice(allTokens, func(i, j int) bool { return allTokens[i] < allTokens[j] })
	totals.TokensP50 = PercentileU64(allTokens, 0.5)
	totals.TokensP95 = PercentileU64(allTokens, 0.95)

	return &ProductivityMetrics{
		Days:   daysOut,
		Totals: totals,
	}, nil
}
